File: src/inkhole/crypto.py
"""
crypto.py
=========
墨洞端到端加密(AES-256-GCM)，被 p2p.py 使用。

整块格式 WHE1(小文件): magic"WHE1" + salt(16B) + nonce(12B) + AES-GCM密文(含校验标签)
分块格式 WHE2/WHE3/WHE4(大文件): magic + salt(16B) + base_nonce(12B) + 帧*
    帧 = [4B 密文长度(BE)] + 密文
    第 i 块: nonce_i = base[0:4] + BE64(BE64(base[4:12]) + i)，AAD = BE64(i)
    块序号进 nonce 和 AAD：篡改、重排、跨文件拼接都会解密失败。
    明文按 4MB 分块——收发内存峰值 4MB，特大文件不再整块进内存。

WHE1/WHE2 为兼容格式，PBKDF2-HMAC-SHA256 使用 10 万次；WHE3 每流 60 万次。
WHE4(经 WHPC "whe4" 能力协商)：口令先经 60 万次 PBKDF2 对固定应用盐派生
主密钥(每进程每口令只算一次并缓存)，每流再用 HKDF-SHA256(master,
salt=流盐, info) 派生流密钥——每次传输省 ~0.5s CPU，流间隔离不变。
salt/nonce 每个文件随机，与 Go whe.go、Android Crypto.kt 逐字节兼容。
"""

from __future__ import annotations
import os
import struct
import threading

_MAGIC = b"WHE1"
_MAGIC2 = b"WHE2"
_MAGIC3 = b"WHE3"
_MAGIC4 = b"WHE4"
_LEGACY_ITERATIONS = 100_000
_ITERATIONS = 600_000
_WHE4_MASTER_SALT = b"INKHOLE-WHE4-MASTER-V1"
_WHE4_STREAM_INFO = b"INKHOLE-WHE4-STREAM-V1"
CHUNK_SIZE = 4 * 1024 * 1024      # 分块明文大小
_CHUNK_OVERHEAD = 20              # 每帧开销: 4B 长度 + 16B GCM tag

_master_lock = threading.Lock()
_master_cache: dict[bytes, bytes] = {}
# 进程随机 HMAC 键：缓存键是口令摘要而非明文口令，内存检查/诊断转储里
# 不会长期留存换掉的旧口令本身(与 Go masterCache 相同的设计)。
_master_cache_key = os.urandom(32)


def _derive_key(secret: str, salt: bytes, iterations: int = _ITERATIONS) -> bytes:
    import hashlib
    return hashlib.pbkdf2_hmac(
        "sha256", secret.encode("utf-8"), salt, iterations, dklen=32)


def _master_cache_id(secret: str) -> bytes:
    import hashlib
    import hmac
    return hmac.new(_master_cache_key, secret.encode("utf-8"),
                    hashlib.sha256).digest()


def _master_key(secret: str) -> bytes:
    """WHE4 主密钥：每口令一次 60 万次 PBKDF2，进程内缓存。"""
    cache_id = _master_cache_id(secret)
    with _master_lock:
        cached = _master_cache.get(cache_id)
    if cached is not None:
        return cached
    derived = _derive_key(secret, _WHE4_MASTER_SALT)
    with _master_lock:
        _master_cache[cache_id] = derived
    return derived


def _stream_key_whe4(secret: str, salt: bytes) -> bytes:
    from cryptography.hazmat.primitives.kdf.hkdf import HKDF
    from cryptography.hazmat.primitives import hashes
    return HKDF(algorithm=hashes.SHA256(), length=32, salt=salt,
                info=_WHE4_STREAM_INFO).derive(_master_key(secret))


# ---------- 整块格式 WHE1 ----------

def encrypt(secret: str, plain: bytes) -> bytes:
    """加密明文 bytes，返回 magic+salt+nonce+密文。"""
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM
    salt, nonce = os.urandom(16), os.urandom(12)
    key = _derive_key(secret, salt, _LEGACY_ITERATIONS)
    return _MAGIC + salt + nonce + AESGCM(key).encrypt(nonce, plain, None)


def decrypt(secret: str, blob: bytes) -> bytes | None:
    """解密；不是加密格式或口令不对返回 None。"""
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM
    from cryptography.exceptions import InvalidTag
    if not is_encrypted(blob) or len(blob) < 4 + 16 + 12 + 16:
        return None
    salt, nonce, ct = blob[4:20], blob[20:32], blob[32:]
    try:
        key = _derive_key(secret, salt, _LEGACY_ITERATIONS)
        return AESGCM(key).decrypt(nonce, ct, None)
    except InvalidTag:
        return None


def is_encrypted(blob: bytes) -> bool:
    """检查数据是否以墨洞整块加密 magic 开头。"""
    return blob[:4] == _MAGIC


# ---------- 分块格式 WHE2/WHE3 ----------

def chunked_wire_size(plain_size: int) -> int:
    """给定明文大小，返回分块加密后的线上字节总数(可预先算出，发送方写进 header)。"""
    n_chunks = (plain_size + CHUNK_SIZE - 1) // CHUNK_SIZE
    return 32 + plain_size + n_chunks * _CHUNK_OVERHEAD


def _chunk_nonce(base: bytes, idx: int) -> bytes:
    """第 idx 块的 nonce：前 4 字节固定，后 8 字节计数器 + idx(mod 2^64)。"""
    ctr = (int.from_bytes(base[4:12], "big") + idx) & 0xFFFFFFFFFFFFFFFF
    return base[:4] + ctr.to_bytes(8, "big")


def encrypt_chunks(secret: str, fileobj, use_whe4: bool = False):
    """流式分块加密生成器：yield 线上字节(先 32B 流头，再逐帧)。

    每次只持有一个 4MB 块，内存峰值恒定。use_whe4 仅当对端在 WHPC 中
    声明了 "whe4" 能力时为 True；否则发 WHE3，任何 v3 对端都能解。
    """
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM
    salt, base = os.urandom(16), os.urandom(12)
    if use_whe4:
        aes = AESGCM(_stream_key_whe4(secret, salt))
        yield _MAGIC4 + salt + base
    else:
        aes = AESGCM(_derive_key(secret, salt))
        yield _MAGIC3 + salt + base
    idx = 0
    while True:
        plain = fileobj.read(CHUNK_SIZE)
        if not plain:
            break
        ct = aes.encrypt(_chunk_nonce(base, idx), plain, struct.pack("!Q", idx))
        yield struct.pack("!I", len(ct)) + ct
        idx += 1


class ChunkedDecryptor:
    """按序解密 WHE2/WHE3/WHE4 分块流。

    用法：先喂 32B 流头构造；随后每读到一帧密文调 decrypt_chunk()，
    返回明文，口令不对/被篡改/被重排返回 None。
    """

    def __init__(self, secret: str, stream_header: bytes):
        from cryptography.hazmat.primitives.ciphers.aead import AESGCM
        if len(stream_header) != 32 or stream_header[:4] not in (
                _MAGIC2, _MAGIC3, _MAGIC4):
            raise ValueError("bad chunked encryption stream header")
        salt = stream_header[4:20]
        if stream_header[:4] == _MAGIC4:
            key = _stream_key_whe4(secret, salt)
        elif stream_header[:4] == _MAGIC2:
            key = _derive_key(secret, salt, _LEGACY_ITERATIONS)
        else:
            key = _derive_key(secret, salt, _ITERATIONS)
        self._base = stream_header[20:32]
        self._aes = AESGCM(key)
        self._idx = 0

    def decrypt_chunk(self, ct: bytes) -> bytes | None:
        from cryptography.exceptions import InvalidTag
        try:
            plain = self._aes.decrypt(_chunk_nonce(self._base, self._idx),
                                      ct, struct.pack("!Q", self._idx))
        except InvalidTag:
            return None
        self._idx += 1
        return plain

File: android/app/src/main/java/com/rexvane/inkhole/p2p/Crypto.kt
package com.rexvane.inkhole.p2p

import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec
import java.nio.ByteBuffer
import java.security.SecureRandom
import java.util.concurrent.ConcurrentHashMap

/**
 * AES-256-GCM 端到端加密, 与桌面版 Python crypto.py、Go whe.go 逐字节兼容。
 *
 * 整块格式 WHE1(小文件): [4B "WHE1"] [16B salt] [12B nonce] [ct + 16B tag]
 * 分块格式 WHE2/WHE3/WHE4(大文件): [4B magic] [16B salt] [12B base_nonce] + 帧*
 *   帧 = [4B 密文长度(BE)] + 密文
 *   第 i 块: nonce_i = base[0:4] + BE64(BE64(base[4:12]) + i)，AAD = BE64(i)
 *   4MB 分块——收发内存峰值恒定，特大文件不再整块进内存。
 * 密钥派生:兼容 WHE1/WHE2 使用 100000 次，WHE3 每流 600000 次。
 * WHE4(经 WHPC "whe4" 能力协商):口令对固定应用盐做一次 600000 次 PBKDF2
 * 得主密钥(进程内缓存)，每流用 HKDF-SHA256(master, salt=流盐, info) 派生
 * 流密钥——每次传输省下秒级派生开销，流间隔离不变。
 */
object Crypto {
    private val MAGIC = "WHE1".toByteArray(Charsets.US_ASCII)
    private val MAGIC2 = "WHE2".toByteArray(Charsets.US_ASCII)
    private val MAGIC3 = "WHE3".toByteArray(Charsets.US_ASCII)
    private val MAGIC4 = "WHE4".toByteArray(Charsets.US_ASCII)
    private val WHE4_MASTER_SALT = "INKHOLE-WHE4-MASTER-V1".toByteArray(Charsets.US_ASCII)
    private val WHE4_STREAM_INFO = "INKHOLE-WHE4-STREAM-V1".toByteArray(Charsets.US_ASCII)
    private const val SALT_LEN = 16
    private const val NONCE_LEN = 12
    private const val LEGACY_ITERATIONS = 100_000
    private const val ITERATIONS = 600_000
    private const val KEY_LEN = 256     // bits
    private const val TAG_LEN = 128     // bits (GCM auth tag)
    const val CHUNK_SIZE = 4 * 1024 * 1024
    private const val CHUNK_OVERHEAD = 20L   // 每帧: 4B 长度 + 16B tag

    private val masterCache = ConcurrentHashMap<String, ByteArray>()
    private const val MASTER_CACHE_MAX = 8  // 进程一般只有一两个口令；上限防病态增长

    // 进程随机 HMAC 键：缓存键是口令摘要而非明文口令，堆转储/诊断里
    // 不会长期留存换掉的旧口令本身(与 Go masterCache 相同的设计)。
    private val masterCacheKey = ByteArray(32).also { SecureRandom().nextBytes(it) }

    private fun masterCacheId(secret: String): String {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(masterCacheKey, "HmacSHA256"))
        return mac.doFinal(secret.toByteArray(Charsets.UTF_8))
            .joinToString("") { "%02x".format(it) }
    }

    /** WHE4 主密钥：每口令一次 60 万次 PBKDF2，进程内缓存。 */
    private fun masterKey(secret: String): ByteArray {
        val cacheId = masterCacheId(secret)
        masterCache[cacheId]?.let { return it }
        val derived = deriveKey(secret, WHE4_MASTER_SALT)
        if (masterCache.size >= MASTER_CACHE_MAX) masterCache.clear()
        masterCache[cacheId] = derived
        return derived
    }

    /** RFC 5869 HKDF-SHA256；32 字节输出恰好一轮 expand。 */
    private fun streamKeyWhe4(secret: String, salt: ByteArray): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(salt, "HmacSHA256"))
        val prk = mac.doFinal(masterKey(secret))
        mac.init(SecretKeySpec(prk, "HmacSHA256"))
        mac.update(WHE4_STREAM_INFO)
        mac.update(1.toByte())
        return mac.doFinal()
    }

    fun isEncrypted(blob: ByteArray): Boolean {
        return blob.size >= 4 + SALT_LEN + NONCE_LEN && blob.copyOfRange(0, 4).contentEquals(MAGIC)
    }

    fun encrypt(secret: String, plain: ByteArray): ByteArray {
        val sr = SecureRandom()
        val salt = ByteArray(SALT_LEN).also { sr.nextBytes(it) }
        val nonce = ByteArray(NONCE_LEN).also { sr.nextBytes(it) }
        val key = deriveKey(secret, salt, LEGACY_ITERATIONS)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(TAG_LEN, nonce))
        val ct = cipher.doFinal(plain)
        return MAGIC + salt + nonce + ct
    }

    fun decrypt(secret: String, blob: ByteArray): ByteArray? {
        if (!isEncrypted(blob)) return null
        val salt = blob.copyOfRange(4, 4 + SALT_LEN)
        val nonce = blob.copyOfRange(4 + SALT_LEN, 4 + SALT_LEN + NONCE_LEN)
        val ct = blob.copyOfRange(4 + SALT_LEN + NONCE_LEN, blob.size)
        val key = deriveKey(secret, salt, LEGACY_ITERATIONS)
        return try {
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(TAG_LEN, nonce))
            cipher.doFinal(ct)
        } catch (e: Exception) {
            null
        }
    }

    private fun deriveKey(secret: String, salt: ByteArray,
                          iterations: Int = ITERATIONS): ByteArray {
        val spec = PBEKeySpec(secret.toCharArray(), salt, iterations, KEY_LEN)
        val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
        return factory.generateSecret(spec).encoded
    }

    // ---------- 分块格式 WHE2/WHE3 ----------

    /** 给定明文大小，返回分块加密后的线上字节总数(发送方写进 header)。 */
    fun chunkedWireSize(plainSize: Long): Long {
        val n = (plainSize + CHUNK_SIZE - 1) / CHUNK_SIZE
        return 32 + plainSize + n * CHUNK_OVERHEAD
    }

    /** 第 idx 块的 nonce：前 4 字节固定，后 8 字节计数器 + idx(mod 2^64，与 Python 一致)。 */
    private fun chunkNonce(base: ByteArray, idx: Long): ByteArray {
        val ctr = ByteBuffer.wrap(base, 4, 8).long + idx   // Long 溢出环绕 == mod 2^64
        return ByteBuffer.allocate(12).put(base, 0, 4).putLong(ctr).array()
    }

    private fun aadOf(idx: Long): ByteArray = ByteBuffer.allocate(8).putLong(idx).array()

    /** 流式分块加密：先取 streamHeader 发出，再逐块 encryptChunk。
     *  useWhe4 仅当对端在 WHPC 中声明 "whe4" 能力时为 true；否则发 WHE3。 */
    class ChunkedEncryptor(secret: String, useWhe4: Boolean = false) {
        private val salt = ByteArray(SALT_LEN).also { SecureRandom().nextBytes(it) }
        private val base = ByteArray(NONCE_LEN).also { SecureRandom().nextBytes(it) }
        private val key = if (useWhe4) streamKeyWhe4(secret, salt) else deriveKey(secret, salt)
        private var idx = 0L
        val streamHeader: ByteArray = (if (useWhe4) MAGIC4 else MAGIC3) + salt + base

        fun encryptChunk(plain: ByteArray, len: Int = plain.size): ByteArray {
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"),
                GCMParameterSpec(TAG_LEN, chunkNonce(base, idx)))
            cipher.updateAAD(aadOf(idx))
            idx++
            return cipher.doFinal(plain, 0, len)
        }
    }

    /** 按序解密 WHE2/WHE3/WHE4 分块流；口令不对/被篡改/被重排返回 null。 */
    class ChunkedDecryptor(secret: String, streamHeader: ByteArray) {
        private val base: ByteArray
        private val key: ByteArray
        private var idx = 0L

        init {
            require(streamHeader.size == 32) { "bad chunked encryption header" }
            val magic = streamHeader.copyOfRange(0, 4)
            val salt = streamHeader.copyOfRange(4, 20)
            base = streamHeader.copyOfRange(20, 32)
            key = when {
                magic.contentEquals(MAGIC4) -> streamKeyWhe4(secret, salt)
                magic.contentEquals(MAGIC3) -> deriveKey(secret, salt, ITERATIONS)
                magic.contentEquals(MAGIC2) -> deriveKey(secret, salt, LEGACY_ITERATIONS)
                else -> throw IllegalArgumentException("bad chunked encryption header")
            }
        }

        fun decryptChunk(ct: ByteArray): ByteArray? = try {
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"),
                GCMParameterSpec(TAG_LEN, chunkNonce(base, idx)))
            cipher.updateAAD(aadOf(idx))
            val plain = cipher.doFinal(ct)
            idx++
            plain
        } catch (e: Exception) {
            null
        }
    }
}
